apps/internal/oauth/ops/authority/authority.go

```diff
@@ -98,6 +98,46 @@ func (r *TenantDiscoveryResponse) Validate() error {
 	return nil
 }
 
+// ValidateIssuerMatchesAuthority validates that the issuer in the TenantDiscoveryResponse matches the authority.
+// This is used to prevent token issuers from impersonating legitimate authorities.
+func (r *TenantDiscoveryResponse) ValidateIssuerMatchesAuthority(authorityURI string) error {
+	if authorityURI == "" {
+		return nil
+	}
+
+	// Parse the issuer URL
+	issuerURL, err := url.Parse(r.Issuer)
+	if err != nil {
+		return fmt.Errorf("TenantDiscoveryResponse: failed to parse issuer URL: %w", err)
+	}
+
+	// Even if it doesn't match the authority, issuers from known and trusted Microsoft hosts are valid
+	if TrustedHost(issuerURL.Host) {
+		return nil
+	}
+
+	// Check against other trusted Microsoft hosts
+	switch issuerURL.Host {
```

Comment:
Can't this check be done after instance discovery? That would be future proof.

Reply:
In the latest commit, it now uses aliases from instance discovery rather than the hardcoded list in the code.